### PR TITLE
Add change attributes and generate thumbnails endpoints

### LINKED ISSUE #83: Add thumbnail endpoint
`GET /orgs/:org/ds/:ds/thumb`

Takes two query parameters `path` and `size` (optional). Only path is required, `size` defaults to 512 if unspecified.

Returns a thumbnail of the specified file (if supported).

On error returns `400` `{"error": "description"}`

See https://github.com/DroneDB/MiniReg/commit/6886bdfd4c856ad8efaa2d563e72c49fa29c4e7b

Requires #75 before being implemented.

### LINKED ISSUE #87: Add chattr endpoint /orgs/:org/ds/:ds/chattr
`POST /orgs/:org/ds/:ds/chattr`

Takes a single form-data parameter `attrs` which is a JSON-encoded string containing the attributes to change:

```
attrs: "{\"public\": true}"
```

Returns the newly updated attributes for the dataset, or 401 `{error: <description>}` otherwise.

See https://github.com/DroneDB/MiniReg/blob/master/libs/dataset.js#L302

Also: https://github.com/DroneDB/DroneDB/blob/master/src/ddb.h#L117

### PR DESCRIPTION
**Pull request by HeDo88TH**
Adds the new `chaddr` and `thumbs` endpoints

Closes #87 
Closes #83

### CODE REVIEW

**Comment by HeDo88TH** [comment]
To test this PR you should put the generated nuget package (1.0.8) in an offline nuget repository in order for the project to compile and run. This will be necessary until we publish the new nuget package on nuget.org

**Comment by pierotofy** [comment]
Looks like the thumbnail endpoint is still missing? Is this ready for review?

**Comment by HeDo88TH** [comment]
No, I'm still refining the workflow on thumbnails side. It will be ready in no time 👍

**Comment by HeDo88TH** [comment]
It's easy to spawn a new Redis instance for testing:

`docker run --name localredis -p 5002:6379 -d redis`

to check cache content

```
docker exec -it localredis redis-cli
dbsize
scan 0
```

I'm writing this down for the next documentation update 👍

**Comment by HeDo88TH** [comment]
This is how the `thumb` endpoint looks in postman

thumb-test

**Comment by pierotofy** [comment]
Nice!! I like the redis approach.

> The cache expiration is set to 5 minutes

A few days (`2` or `3`) might be better as a default?

The `thumb` endpoint is almost correct, the parameters `path` and `size` should be part of the URL query (thumb?path=...&size=...).

**Review by pierotofy — COMMENTED** [review]
`thumb` endpoint needs to match MiniReg's spec.

**Comment by HeDo88TH** [comment]
Now it is ok!

File: Registry.Web/Utilities/Extenders.cs [hunk 1]
```diff
@@ -178,6 +181,17 @@ public static TagDto ToTag(this string tag)
...
         }
 
+        public static T ToObject<T>(this JsonElement obj)
+        {
+            return JsonConvert.DeserializeObject<T>(obj.GetRawText());
+        }
+
+        public static T ToObject<T>(this Dictionary<string, object> obj)
+        {
+            // Just don't ask please
```
Comment (pierotofy) [inline Registry.Web/Utilities/Extenders.cs:191]: lol

image

https://www.destroyallsoftware.com/talks/wat

Comment (HeDo88TH) [reply Registry.Web/Utilities/Extenders.cs:191]: I'm looking for a more elegant solution but, for now, this is the best I found at 2am

**Comment by pierotofy** [comment]
Looks good. I've found some issues with other endpoints, but can be fixed as part of another PR. :+1: 

Thanks!

**Pull request merged**